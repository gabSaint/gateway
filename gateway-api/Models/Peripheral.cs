using gateway_api.Contexts;
using gateway_api.Extensions;
using gateway_api.Models.Validations;
using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Linq;
using System.Threading.Tasks;

namespace gateway_api.Models
{
    public class Peripheral
    {
        [Key]
        public int Id { get; set; }

        [Required]
        public string Vendor { get; set; }

        [Required]
        public string Date { get; set; }

        [Required]
        [ValidStatus]
        public string Status { get; set; }

        [Required]
        [ForeignKey("Gateway")]
        public int GatewayId { get; set; } 

    }

    public static class PeripheralSeed
    {
        public static void InitData(DataContext context)
        {
            var rnd = new Random();

            var vendors = new[] { "ASL", "AFG", "HJP", "GER", "SPI" };

            var statuss = new[] { "online", "offline" };

            var dates = new[] { "Mon Jan 11 2021", "Tue Jan 12 2021", "Wed Jan 13 2021", "Thu Jan 14 2021", "Fri Jan 15 2021"  };

            context.Peripherals.AddRange(100.Times(x =>
            {
                var vendor = vendors[rnd.Next(0, 5)];
                var status = statuss[rnd.Next(0, 2)];
                var date = dates[rnd.Next(0, 5)];
                var id = $"{x,-3:000}";
                var gatewayId = rnd.Next(0, 20);

                return new Peripheral
                {
                    Id = x,
                    Vendor = vendor,
                    Date = date,
                    Status = status,
                    GatewayId = gatewayId
                };
            }));

            context.SaveChanges();
        }
    }
}
