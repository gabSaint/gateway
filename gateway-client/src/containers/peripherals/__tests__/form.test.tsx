import React from "react";
import { ReactWrapper, shallow, ShallowWrapper } from "enzyme";
import renderer from "react-test-renderer";

import PeripheralForm, {
  FormContent,
  Props as PeripheralProps,
  validate,
} from "../form";
import Peripheral from "models/peripheral";
import { FormikState } from "formik";

describe("PeripheralForm", () => {
  let wrapperPeripheralForm: ShallowWrapper | ReactWrapper;
  let props: PeripheralProps;

  beforeEach(() => {
    props = {
      peripheral: {} as Peripheral,
      handleSubmit: jest.fn(),
    };
    wrapperPeripheralForm = shallow(<PeripheralForm {...props} />);
  });

  it("renders correctly", () => {
    const tree = renderer.create(<PeripheralForm {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders form Formik", () => {
    const formik = wrapperPeripheralForm.find("Formik");
    expect(formik).toHaveLength(1);
  });
});

describe("FormContent", () => {
  let wrapperPeripheralFormContent: ReactWrapper | ShallowWrapper;
  let props: FormikState<Peripheral>;

  beforeEach(() => {
    props = {
      errors: { id: "error", vendor: "error", status: "error" },
      touched: { id: true, vendor: true, status: true },
      values: { id: 123, vendor: "ATS", status: "online" },
    } as FormikState<Peripheral>;
    wrapperPeripheralFormContent = shallow(<FormContent {...props} />);
  });

  describe("Form", () => {
    it("renders submit button", () => {
      const buttonSubmit = wrapperPeripheralFormContent.find(
        "[data-test='button-submit']"
      );
      expect(buttonSubmit).toHaveLength(1);
    });

    it("renders cancel button", () => {
      const buttoncancel = wrapperPeripheralFormContent.find(
        "[data-test='button-cancel']"
      );
      expect(buttoncancel).toHaveLength(1);
    });

    it("renders vendor FormInput", () => {
      const inputVendors = wrapperPeripheralFormContent.find(
        "FormInput[name='vendor']"
      );
      expect(inputVendors).toHaveLength(1);
    });

    it("renders status FormRadio", () => {
      const inputStatus = wrapperPeripheralFormContent.find(
        "FormRadio[label='Status']"
      );
      expect(inputStatus).toHaveLength(1);
    });

    it("renders 2 RadioInput", () => {
      const inputRadios = wrapperPeripheralFormContent.find("RadioInput");
      expect(inputRadios).toHaveLength(2);
    });
  });

  describe("Function Validate", () => {
    let correctPeripheral: any;
    let wrongPeripheral: any;

    beforeEach(() => {
      correctPeripheral = validate(
        new Peripheral(1234, "AID", "24/1/2020", "online", 1)
      );
      wrongPeripheral = validate({} as Peripheral);
    });

    it("validates correctly the vendor input", () => {
      expect(correctPeripheral.vendor).toBeUndefined();
      expect(wrongPeripheral.vendor).toBeDefined();
    });

    it("validates correctly the status input", () => {
      expect(correctPeripheral.status).toBeUndefined();
      expect(wrongPeripheral.status).toBeDefined();
    });
  });
});
