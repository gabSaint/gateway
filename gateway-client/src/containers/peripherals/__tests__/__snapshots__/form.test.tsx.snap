// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PeripheralForm renders correctly 1`] = `
<form
  action="#"
  className="ui form"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="field"
  >
    <label
      htmlFor="input-vendor"
    >
      Vendor
    </label>
    <input
      id="input-vendor"
      name="vendor"
      onBlur={[Function]}
      onChange={[Function]}
      value=""
    />
  </div>
  <div
    className="inline fields"
    role="group"
  >
    <label>
      Status
    </label>
    <div
      className="field"
    >
      <div
        className="ui radio checkbox"
      >
        <input
          checked={false}
          className="radio-item"
          name="status"
          onBlur={[Function]}
          onChange={[Function]}
          tabIndex="0"
          type="radio"
          value="online"
        />
        <label>
          Online
        </label>
      </div>
    </div>
    <div
      className="field"
    >
      <div
        className="ui radio checkbox"
      >
        <input
          checked={false}
          className="radio-item"
          name="status"
          onBlur={[Function]}
          onChange={[Function]}
          tabIndex="0"
          type="radio"
          value="offline"
        />
        <label>
          Offline
        </label>
      </div>
    </div>
  </div>
  <button
    data-test="button-submit"
    type="submit"
  >
    Submit
  </button>
  <div>
    <button
      data-test="button-cancel"
      onClick={[Function]}
      type="button"
    >
      Cancel
    </button>
  </div>
</form>
`;
